libs/ml/include/ml/optimisation/lazy_adam_optimiser.hpp

```diff
@@ -0,0 +1,371 @@
+  fetch::math::Pow(beta1_, static_cast<DataType>(this->epoch_ + 1), beta1_t_);
+  fetch::math::Pow(beta2_, static_cast<DataType>(this->epoch_ + 1), beta2_t_);
+
+  std::vector<SizeSet> rows;
+
+  while (gradient_it != this->gradients_.end())
+  {
+    // Skip frozen trainables
+    if ((*trainable_it)->GetFrozenState())
+    {
+      continue;
+    }
+
+    std::pair<TensorType const, SizeSet const> gradient_pair =
+        (*trainable_it)->GetSparseGradientsReferences();
+    rows.push_back(gradient_pair.second);
+
+    // Normal ApplyGradient
+    // if number_of_rows_to_update*sparsity_threshold_ > total_rows
+    if (rows.at(rows.size() - 1).empty() ||
+        (rows.at(rows.size() - 1).size() * sparsity_threshold_) > gradient_pair.first.shape().at(1))
+    {
+      ApplyLogic(batch_size, *gradient_it, *momentum_it, *mt_it, *vt_it, *cached_weight_it,
+                 (*trainable_it)->GetGradientsReferences());
+    }
+
+    else
+    {
+      // Sparse apply gradient
+      // if number_of_rows_to_update*sparsity_threshold_ <= total_rows
+
+      for (SizeType update_index : rows.at(rows.size() - 1))
+      {
+
+        auto       gradient_slice        = gradient_it->Slice(update_index, 1);
+        TensorType gradient_slice_tensor = gradient_slice.Copy();
+
+        auto       momentum_slice        = momentum_it->Slice(update_index, 1);
+        TensorType momentum_slice_tensor = momentum_slice.Copy();
+
+        auto       mt_slice        = mt_it->Slice(update_index, 1);
+        TensorType mt_slice_tensor = mt_slice.Copy();
+
+        auto       v_slice        = vt_it->Slice(update_index, 1);
+        TensorType v_slice_tensor = v_slice.Copy();
+
+        auto       cache_slice        = cached_weight_it->Slice(update_index, 1);
+        TensorType cache_slice_tensor = cache_slice.Copy();
+
+        auto       refs_slice = (*trainable_it)->GetGradientsReferences().Slice(update_index, 1);
+        TensorType refs_slice_tensor = refs_slice.Copy();
+
+        ApplyLogic(batch_size, gradient_slice_tensor, momentum_slice_tensor, mt_slice_tensor,
+                   v_slice_tensor, cache_slice_tensor, refs_slice_tensor);
+
+        // Put things back
+        gradient_slice.Assign(gradient_slice_tensor);
+        momentum_slice.Assign(momentum_slice_tensor);
+        mt_slice.Assign(mt_slice_tensor);
+        v_slice.Assign(v_slice_tensor);
+        cache_slice.Assign(cache_slice_tensor);
+      }
+
+      // we need to explicitly reset the gradients for this shared op to avoid double counting
+      // in the case of shared ops
+      (*trainable_it)->ResetGradients();
+    }
+
+    ++cached_weight_it;
+    ++momentum_it;
+    ++mt_it;
+    ++vt_it;
+
+    ++gradient_it;
+    ++trainable_it;
+  }
+
+  // calling apply gradients on the graph ensures that the node caches are reset properly
+  this->graph_->ApplySparseGradients(this->gradients_, rows);
+}
+
+template <class T>
+void LazyAdamOptimiser<T>::ResetCache()
```

Comment:
seems like this method would be redundant if you inherit from AdamOptimiser instead of Optimiser